app/models/follow_request.rb

```diff
@@ -25,6 +25,17 @@ class FollowRequest < ApplicationRecord
 
   enum status: { pending: "pending", rejected: "rejected", accepted: "accepted" }
 
+  # Automatic scopes from enum :status
   # scope :accepted, -> { where(status: "accepted" ) }
   # scope :not_accepted, -> { where.not(status: "accepted" ) }
+
+  validates :recipient_id, uniqueness: { scope: :sender_id, message: "already requested" }
+
+  validate :users_cant_follow_themselves
+
+  def users_cant_follow_themselves
+    if sender_id == recipient_id
+      errors.add(:sender_id, "can't follow themselves")
```

Comment:
Should this error message be

```suggestion
      errors.add(:sender_id, "can't follow yourself")
```

Since this would only happen if the current user tries to follow their own account?

Reply:
Until we get into customizing error messages better using `I18N`, we're limited to appending the message onto `column_name.humanized`, so your suggestion would end up being "sender can't follow yourself" - the current version would be "sender can't follow themselves". The latter isn't great but I think it's better.

I also considered variants that start with "Recipient..." but couldn't come up with anything great. I figured we'd use the clunkiness it as a _lock_ to motivate learning about I18n. Thoughts?

Reply:
Ah, that's right. I still mixup how the error messages are structured.

Maybe something like "sender can't follow their own account", is more clear?

I also think their is value in this being a lock before key situation.

Reply:
In the `errors.add` call you include `:sender_id` yet my understanding of this discussion is that it will only append the message to `column_name.humanized`. Is the addition of `:sender_id` to enable future customization using `I18N`?

Reply:
> In the `errors.add` call you include `:sender_id` yet my understanding of this discussion is that it will only append the message to `column_name.humanized`. Is the addition of `:sender_id` to enable future customization using `I18N`?

We have two options — we can either attach the validation error to:

 - one of the object's attributes, e.g.:

    ```ruby
    errors.add(:sender_id, "can't follow themselves")
    ```
 - or `:base`, e.g.:

    ```ruby
    errors.add(:base, "You can't follow yourself")
    ```

The latter is supposed to be used for errors that relate to the object's state as a whole, rather than any individual column; and now that we are having this discussion, I think actually that might be the right choice here, since this error relates to the combination of `sender_id` and `recipient_id` and that is all that there is to a `FollowRequest`.

In any case, the reason to not _always_ use `:base` is because if an error does only relate to a single attribute, it's better to be specific and attach it to that column. For one thing Rails will humanize the column name and append it to the error message automatically, which is sometimes handy but sometimes it is constraining, as in this case. More importantly, we can design better forms and put valid/invalid feedback right on each individual input if we have attribute-level errors, as you can observe in the Devise sign up/edit profile forms.